linkerd/meshtls/boring/src/client.rs

```diff
@@ -76,25 +81,36 @@ where
             let conn = connector.map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
             let config = conn
                 .configure()
-                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
-            let io = tokio_boring::connect(config, server_name.as_str(), io)
+                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?
+                // Switch off DNS SAN Verification based on the hostname
+                .verify_hostname(false);
+
+            tokio_boring::connect(config, server_name.as_str(), io)
                 .await
                 .map_err(|e| match e.as_io_error() {
                     // TODO(ver) boring should let us take ownership of the error directly.
                     Some(ioe) => io::Error::new(ioe.kind(), ioe.to_string()),
                     // XXX(ver) to use the boring error directly here we have to constraint the socket on Sync +
                     // std::fmt::Debug, which is a pain.
                     None => io::Error::new(io::ErrorKind::Other, "unexpected TLS handshake error"),
-                })?;
-
-            debug!(
-                tls = io.ssl().version_str(),
-                client.cert = ?io.ssl().certificate().and_then(super::fingerprint),
-                peer.cert = ?io.ssl().peer_certificate().as_deref().and_then(super::fingerprint),
-                alpn = ?io.ssl().selected_alpn_protocol(),
-                "Initiated TLS connection"
-            );
-            Ok(ClientIo(io))
+                }).and_then(|io| {
+                    let cert = io.ssl()
+                        .peer_certificate()
+                        .ok_or_else(|| io::Error::new(
+                            io::ErrorKind::Other,
+                            "could not extract peer cert",
+                        ))?;
+                    verify::verify_id(&cert, &server_id)
+                        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+                    debug!(
+                        tls = io.ssl().version_str(),
+                        client.cert = ?io.ssl().certificate().and_then(super::fingerprint),
+                        peer.cert = ?io.ssl().peer_certificate().as_deref().and_then(super::fingerprint),
+                        alpn = ?io.ssl().selected_alpn_protocol(),
+                        "Initiated TLS connection"
+                    );
+                    Ok(ClientIo(io))
+                })
```

Comment:
I would generally advise against using chained calls (i.e. and_then) if they don't add clarity and needlessly nest logic. I also think this verification logic warrants a clearer overview here.

It's also seems appropriate to log the server name and id at trace level, at least, so the values can be inspected in a pinch.

Suggested as:

```rust
        Box::pin(async move {
            let config = connector
                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?
                .configure()
                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;


            // Establish a TLS connection to the server using the provided
            // `server_name` as an SNI value to the server.
            //
            // Hostname verification is DISABLED, as we do not require that the
            // peer's certificate actually matches the `server_name`. Instead,
            // the `server_id` is used to perform the appropriate form of
            // verification after the session is established.
            let io = tokio_boring::connect(config.verify_hostname(false), server_name.as_str(), io)
                .await
                .map_err(|e| match e.as_io_error() {
                    // TODO(ver) boring should let us take ownership of the error directly.
                    Some(ioe) => io::Error::new(ioe.kind(), ioe.to_string()),
                    // XXX(ver) to use the boring error directly here we have to
                    // constrain the socket on Sync + std::fmt::Debug, which is
                    // a pain.
                    None => io::Error::new(io::ErrorKind::Other, "unexpected TLS handshake error"),
                })?;

            // Servers must present a peer certificate. We extract the x509 cert
            // and verify it manually against the `server_id`.
            let cert = io.ssl().peer_certificate().ok_or_else(|| {
                io::Error::new(io::ErrorKind::Other, "could not extract peer cert")
            })?;
            verify::verify_id(&cert, &server_id)
                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;

            debug!(
                tls = io.ssl().version_str(),
                client.cert = ?io.ssl().certificate().and_then(super::fingerprint),
                peer.cert = ?io.ssl().peer_certificate().as_deref().and_then(super::fingerprint),
                alpn = ?io.ssl().selected_alpn_protocol(),
                "Initiated TLS connection"
            );
            trace!(peer.id = %server_id, peer.name = %server_name)
            Ok(ClientIo(io))
        })
```